Add background info by garnetgrimm to faq

krystian-hebel opened the pull request:
Continuation of https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41, based on original submission to https://github.com/TrenchBoot/documentation/pull/11. Remaining issues are related to intermediate loader, and it would be best if @dpsmith could look at those. I'll add links to original discussion to the related lines here.

philipanda committed:
FAQ.md: Add reformatted Background.md contents

Originally submitted to https://github.com/TrenchBoot/documentation/pull/11

Co-authored-by: Garnet Grimm
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

philipanda committed:
docs/FAQ.md: remove outdated info, fix factual errors

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented on docs/FAQ.md:
```diff
@@ -1,15 +1,126 @@
+and [Multiboot2 UEFI](https://github.com/TrenchBoot/trenchboot-issues/milestone/11)
+will be added later.
+
+### Intermediate Phase - Linux Kernel / U-ROOT (Initramfs)
+
+The intermediate phase processes information collected in the bootstrap phase
+and converts it into normalized data. This data is eventually used to determine
+whether to boot into the target OS. The kernel and u-root initramfs together
+make up TrenchBoot's _Security Engine_, which is an intermediate mini-OS that
+serves to process data gathered by that bootstrap phase. During the
+intermediate phase, the kernel and initramfs work together to measure block
+devices, individual files, SMBUS/DMI information and provides unseal based,
+external device and network based attestation. TrenchBoot calls this
+functionality secure launch (aka slaunch).
```
https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41#discussion_r1739735720 (reference to `table in section 4` already removed).

krystian-hebel commented on docs/FAQ.md:
```diff
@@ -1,15 +1,126 @@
+external device and network based attestation. TrenchBoot calls this
+functionality secure launch (aka slaunch).
+
+## 3. Why does TrenchBoot use an intermediate launcher?
```
https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41#discussion_r1739744306

dpsmith replied:
The intermediate loader/launcher is just one use case, it just happens to be the use case that was implemented because, 1. it effectively mimics how tboot works allowing it to be a drop-in replacement, 2. it met the need of Oracle, and 3. provides components that can be reused to implement other use cases.

If this question is to stay, then it should read "What is the TrenchBoot intermediate loader?", and the answer should be updated that it tends to be the most common DRTM use-case, it's how tboot works and how Microsoft's Secure Core DRTM works.

krystian-hebel replied:
@dpsmith https://github.com/TrenchBoot/TrenchBoot.github.io/pull/45/commits/d2c5d71d0c6f66d06b5d3ae94754a8b5b700c3f5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R122-R138 WDYT?

krystian-hebel commented on docs/FAQ.md:
```diff
@@ -36,8 +146,7 @@ a strong assertion to correctness that can be repeatedly extended and verified.
 
-
-## 3. What do I need to incorporate TrenchBoot into my system? 
+## 5. What do I need to incorporate TrenchBoot into my system?
```
https://github.com/TrenchBoot/TrenchBoot.github.io/pull/41#discussion_r1739744796

dpsmith commented on docs/FAQ.md:
```diff
@@ -18,10 +129,9 @@ intermediary loaders with a common Linux-based loader that provides a rich
 
-
-## 2. What are the benefits of measurement over signature validation?
+## 4. What are the benefits of measurement over signature validation?
```
I would not say "What are the benefits of measurement over signature validation?", I would say "How do measurement trust chains contrast with verification trust chains?".

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
-are two primary schools of thought: Static Root of Trust Measurements and
-Dynamic Root of Trust Measurements.
+Like our [Glossary](./theory/Glossary.md#trust) says,
+trust is an assured reliance on the properties, ability, strength, or truth of
```
Rearrange this sentence a bit for brevity:
"Trust is an assured reliance on the properties, ability, strength, or truth of an entity (See [Glossary](./theory/Glossary.md#trust))."

Edited-by: Amanda Ferguson <af.jenks@gmail.com>

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+and proving it, we need to trust it. This might include the belief, that our
+computer will behave in an expected way, that it won't leak our data, that it
+won't do something we didn't intend to do.
+
```
Minor changes here:
"This might include the belief that our computer will behave in an expected way, that it won't leak our data, and that it won't do something we didn't intend or expect."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+computer will behave in an expected way, that it won't leak our data, that it
+won't do something we didn't intend to do.
+
+When establishing the trust of system components there are two primary schools
```
Comma after components

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
-boot block and then moves on to measure the rest of the bios.
+SRTM Starts with Core Root of Trust for Measurement (CRTM), and works its way
+up the chain where each link verifies the next. In other words, SRTM starts by
+measuring the BIOS boot block and then moves on to measure the rest of the BIOS.
```
Remove "then"

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+SRTM has a few problems:
+
+- Large TCB (Trusted Computing Base) - the more software as a part of the chain
+of trust, the more potential vectors of attack.
```
"- Large TCB (Trusted Computing Base): the more software added to the chain of trust, the more potential vectors of attack."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+
+- Large TCB (Trusted Computing Base) - the more software as a part of the chain
+of trust, the more potential vectors of attack.
+- Takes place before boot - reassuring the validity of the chain of trust
```
Put a : after boot like previous item.

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+- Takes place before boot - reassuring the validity of the chain of trust
+requires a reboot.
+- Only provides load-time integrity assurance - the integrity was maintained
+during boot, but there is no guarantee it will stay like that.
```
"- Only provides load-time integrity assurance: integrity is maintained during boot, but there is no guarantee it will remain so."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
 
 ### DRTM
 
 DRTM aims to address shortcomings of SRTM. DRTM creates trusted environment from
```
"creates a trusted..."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,53 +1,57 @@
+dynamic launch, can impact software started after the dynamic launch. For this
+reason, DRTM takes place after boot, as opposed to SRTM, which takes place
+pre-boot. Intel's solution for DRTM is TXT, which utilizes special instructions
+on the processor (SMX). AMD's solution for DRTM is called SKINIT. TrenchBoot
```
"Overall, TrenchBoot provides a solution for setting up DRTM on an arbitrary platform."

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+[txt-spec]: https://www.intel.com/content/www/us/en/content-details/315168/intel-trusted-execution-technology-intel-txt-software-development-guide.html
+[skinit-spec]: https://www.amd.com/system/files/TechDocs/24593.pdf
+
+### So, why TrenchBoot?
```
Drop the So - just Why TrenchBoot

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -68,7 +72,11 @@ The process for carrying out DTRM depends on your system:
 - Most DRTM systems support only one attestation action
 
-## 2. How does TrenchBoot work? (trenchboot architecture)
+Trenchboot is meant to solve these, and many other problems by providing a
```
Remove comma

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+[the corresponding page](./theory/Architecture.md).*
+![Overview](img/Architectural_Flow.png)
+This diagram shows a general overview of the overall architecture of a
+TrenchBoot system. In Bootstrap phase the system environment is measured and
```
Add a comma after "Bootstrap phase" and remove the comma after PCRs in the next line.

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+TrenchBoot system. In Bootstrap phase the system environment is measured and
+that information stored in the dynamic PCRs, using whatever method is
+appropriate for the given hardware. The intermediate phase analyzes the
+information collected in the bootstrap phase to decide if it wants to continue
```
Change "if it wants to" to "whether" and remove comma after OS on next line.

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+information collected in the bootstrap phase to decide if it wants to continue
+booting and pass control to the target OS, or if it wants to stop immediately.
+If desired, additional evidence/measurements can be gathered in this phase. Once
+ the integrity of the system is verified the TrenchBoot kernel passes off
```
is verified,

rossphilipson commented on docs/FAQ.md:
```diff
@@ -82,78 +90,37 @@ appropriate for the given hardware. The intermediate phase analyzes the
-(although it's available on a separate branch).
+TrenchBoot utilizes its own system to carry out TXT on Intel, or SKINIT on AMD,
+which is a custom version of GRUB that has commands built in to carry out a
+TrenchBoot _Secure Launch_. Those commands are `slaunch` and `slaunch_module`.
```
Change to:
"GRUB has commands built in to carry out a TrenchBoot _Secure Launch_ (currently supporting Intel TXT and AMD SKINIT). This is typically called the pre-launch or preamble phase of the launch. These commands are `slaunch` and `slaunch_module`."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -82,78 +90,37 @@ appropriate for the given hardware. The intermediate phase analyzes the
+
+The custom GRUB works with:
+
+- Linux boot protocol (Intel only, legacy and UEFI),
```
Remove end punctuation on list items to be consistent with other lists

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -82,78 +90,37 @@ appropriate for the given hardware. The intermediate phase analyzes the
-whether to boot into the target OS. The kernel and u-root initramfs
-together make up Trenchboot's "Security Engine", which is an intermediate
-mini-OS that serves to process data gathered by that bootstrap phase. During the
+whether to boot into the target OS. The kernel and u-root initramfs together
```
Reword sentence:
"Together the kernel and u-root initramfs make up TrenchBoot's _Security Engine_, an intermediate mini-OS that processes data gathered by that bootstrap phase."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -82,78 +90,37 @@ appropriate for the given hardware. The intermediate phase analyzes the
+make up TrenchBoot's _Security Engine_, which is an intermediate mini-OS that
+serves to process data gathered by that bootstrap phase. During the
 intermediate phase, the kernel and initramfs work together to measure block
 devices, individual files, SMBUS/DMI information and provides unseal based,
```
Change to "information, etc. The kernel and initramfs also provide unseal..."

rossphilipson commented on docs/FAQ.md:
```diff
@@ -82,78 +90,37 @@ appropriate for the given hardware. The intermediate phase analyzes the
-
-[acm-download]: https://www.intel.com/content/www/us/en/developer/articles/tool/intel-trusted-execution-technology.html
+external device and network based attestation. TrenchBoot calls this
+functionality secure launch (aka slaunch).
```
s/secure launch/Secure Launch

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+
+## 4. Why does TrenchBoot use an intermediate launcher?
 
 For Linux systems doing both verified(secure) and measured boot, there is an
```
Missing space after verified

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

krystian-hebel replied:
https://github.com/TrenchBoot/TrenchBoot.github.io/blob/cleanup/docs/FAQ.md?plain=1#L14, should fix itself after rebase.

rossphilipson commented on docs/FAQ.md:
```diff
@@ -1,15 +1,159 @@
+## 4. Why does TrenchBoot use an intermediate launcher?
 
 For Linux systems doing both verified(secure) and measured boot, there is an
 intermediary that handles the security enforcement. For verified boot it is the
```
"For verified boot, the intermediary is the UEFI shim loader and for measured boot, it is tboot."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -18,10 +162,9 @@ intermediary loaders with a common Linux-based loader that provides a rich
 security processing framework. One role that TrenchBoot does not fulfill is
```
Starting with the sentence "One role...", this section needs some rework and clarification. We were not exactly sure what that should look like because we did not quite understand what was trying to be conveyed here.

Ross

krystian-hebel replied:
This paragraph has been reworked in https://github.com/TrenchBoot/TrenchBoot.github.io/pull/45/commits/d2c5d71d0c6f66d06b5d3ae94754a8b5b700c3f5#diff-0bb42a85d04c8b7c9b659307979fd43e8f6e084283a300bfff959adfcbddfd17R122-R138

rossphilipson commented on docs/FAQ.md:
```diff
@@ -46,11 +188,27 @@ engine components you desire, provide any necessary configurations, and build
 an instance of the loader. After that, you configure your system boot to launch
```
s/After that, you/Next,/

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -46,11 +188,27 @@ engine components you desire, provide any necessary configurations, and build
-
-The [TrenchBoot Blueprints](
-    https://github.com/TrenchBoot/documentation/tree/master/blueprints)
+The [TrenchBoot Blueprints](./blueprints/index.md)
```
Reword:
"[TrenchBoot Blueprints](./blueprints/index.md) collect product feature requests. Check to see if there is an existing blueprint that addresses your feature request. You may also submit a blueprint via pull request to suggest features for implementation."

Edited-by: Amanda Ferguson [af.jenks@gmail.com](mailto:af.jenks@gmail.com)

rossphilipson commented on docs/FAQ.md:
```diff
@@ -36,8 +179,7 @@ a strong assertion to correctness that can be repeatedly extended and verified.
 It therefore relies on the ability to know what correct is and to securely
 verify measurement with expected correctness.
```
The previous section could use some grammatical and sentence structure work but it is not really part of this PR. Maybe it can be fixed up in a subsequent PR after all the work here is in.

philipanda committed:
FAQ.md: Language and interpunction

Edited-by: Amanda Ferguson <af.jenks@gmail.com>
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
docs/FAQ.md: Explain why TB uses an intermediate loader

docs/FAQ.md carry-out -> carry out
docs/FAQ.md: Change Section 3 title
docs/FAQ.md: interpunction
docs/FAQ.md: Remove "The" from TrenchBoot Blueprints ref

Co-authored-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

rossphilipson commented:
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
